### PR TITLE
EventParser Implementation

### LINKED ISSUE #20430: Event Parser for distributed nexus loader
Part of #20344 

This issue deals with the creation of an object which parses event data, prepares for sending on different mpi ranks, re-distributes data accross ranks and fills some output object (EventLists?).

### PR DESCRIPTION
**Pull request by LamarMoore**
Implementation of the event parser for parallel event loading/processing. See #20344. 

**To test:**

Core review.

Fixes #20430. 

Internal change, no release notes.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Parallel/src/IO/EventParser.cpp
Comment (SimonHeybrock) [inline Framework/Parallel/src/IO/EventParser.cpp]: Does it make sense to pass by value and move?

File: Framework/Parallel/inc/MantidParallel/IO/EventParser.h
Comment (SimonHeybrock) [inline Framework/Parallel/inc/MantidParallel/IO/EventParser.h]: Why do these need to be mutable?

File: Framework/Parallel/inc/MantidParallel/IO/EventParser.h
Comment (SimonHeybrock) [inline Framework/Parallel/inc/MantidParallel/IO/EventParser.h]: Why const?

File: Framework/Parallel/src/IO/EventParser.cpp
Comment (SimonHeybrock) [inline Framework/Parallel/src/IO/EventParser.cpp]: Could do this in-place in `event_id` to save memory bandwidth?

File: Framework/Parallel/src/IO/EventParser.cpp
Comment (SimonHeybrock) [inline Framework/Parallel/src/IO/EventParser.cpp]: I think I would do the translation into `TofEvent` on the receiver side. `TofEvent` has 16 bytes (aligned to 8 byte), so that brings `Event` to 24 bytes. If you keep the original types it might be only 16 byte (if `event_time_offset` is 4 byte). Also it may be the case that we use not all ranks for loading, so as much work as possible should be done on the receiver side.

**Review by SimonHeybrock — COMMENTED** [review]
Just a quick review pass before lunch, have not made it to the tests yet. Structure looks mostly good, I think:

File: Framework/Parallel/src/IO/EventParser.cpp
Comment (SimonHeybrock) [inline Framework/Parallel/src/IO/EventParser.cpp]: I am not sure such a check makes sense. You will be receiving input data with arbitrary `event_id` (or `bankIndex`), and unless you check every single one you have no sensible way of checking the length of `eventLists` (or `bankOffsets`). Having these to checks here gives a false sense of security, and having length 0 is no worse than having the length off by 1.

Maybe it makes more sense to put verbose warning comments at the appropriate points, saying that there is no range check, i.e., you just have to rely on the client using this class.

**Comment by LamarMoore** [comment]
@SimonHeybrock @OwenArnold This represents a first pass of an MPI implementation of event parsing. In this implementation, data is redistributed and then written to the EventLists. I will interleave this operation in the next iteration if necessary, as such, I have not had any use for the `RankGroups`.

File: Framework/Parallel/src/IO/EventParser.cpp
Comment (OwenArnold) [inline Framework/Parallel/src/IO/EventParser.cpp]: If this `data` param represents data on other ranks, presumably we need to indicate that it cannot be modified on the current rank. i.e. should be ```const std::vector<std::vector<Event>>&```?

File: Framework/Parallel/test/EventParserTest.h
Comment (OwenArnold) [inline Framework/Parallel/test/EventParserTest.h]: Commented out block?

File: Framework/Parallel/test/EventParserTest.h
Comment (OwenArnold) [inline Framework/Parallel/test/EventParserTest.h]: Naming is good. But would be nice to have brief comments somewhere on how this is generating the events for use by the parser. Looks like there is quite a bit of logic in this type.

File: Framework/Parallel/test/EventParserTest.h
Comment (OwenArnold) [inline Framework/Parallel/test/EventParserTest.h]: Don't understand why this creational method is on the generator. Doesn't make 100% sense to me.

File: Framework/Parallel/test/EventParserTest.h
Comment (OwenArnold) [inline Framework/Parallel/test/EventParserTest.h]: Are you going to add performance tests as part of this PR?

**Review by OwenArnold — CHANGES_REQUESTED** [review]
Starting to look good. I think we should have a conversation about some aspects of this PR.

File: Framework/Parallel/test/EventParserTest.h
Comment (LamarMoore) [reply Framework/Parallel/test/EventParserTest.h]: Yes, currently writing those.

File: Framework/Parallel/test/EventParserTest.h
Comment (LamarMoore) [reply Framework/Parallel/test/EventParserTest.h]: That should have been removed.

File: Framework/Parallel/inc/MantidParallel/IO/EventParser.h
Comment (LamarMoore) [reply Framework/Parallel/inc/MantidParallel/IO/EventParser.h]: Yes. I need to rebase after another of Simon's PRs.

File: Framework/Parallel/inc/MantidParallel/IO/EventParser.h
Comment (LamarMoore) [reply Framework/Parallel/inc/MantidParallel/IO/EventParser.h]: 👍

File: Framework/Parallel/src/IO/EventParser.cpp
Comment (LamarMoore) [reply Framework/Parallel/src/IO/EventParser.cpp]: event_ids are detector ids. I will update the comment

File: Framework/Parallel/src/IO/EventParser.cpp
Comment (LamarMoore) [reply Framework/Parallel/src/IO/EventParser.cpp]: The comment below shows **NB event_id_start is transformed to contain the spectrum indices** as well as the documentation for `event_id`. I will try to use better wording.

File: Framework/Parallel/test/EventParserTest.h
Comment (LamarMoore) [reply Framework/Parallel/test/EventParserTest.h]: I did this so I could access these methods from the performance test. My initial idea was to create a parser based on the data generated. This will be modified shortly either way.

File: Framework/Parallel/inc/MantidParallel/IO/EventParser.h
Comment (OwenArnold) [inline Framework/Parallel/inc/MantidParallel/IO/EventParser.h]: :thumbsup:

File: Framework/Parallel/src/IO/EventParser.cpp
Comment (OwenArnold) [inline Framework/Parallel/src/IO/EventParser.cpp]: Not sure we need `finalize` at present.

File: Framework/Parallel/src/IO/EventParser.cpp
Comment (OwenArnold) [inline Framework/Parallel/src/IO/EventParser.cpp]: I think it would be useful to know that what is started is done in an asynchronous manner, as you need to know this for the other methods to make sense. Can you make this `startAsync`  or `executeAsync`?

**Review by OwenArnold — CHANGES_REQUESTED** [review]

File: Framework/Parallel/test/EventParserTest.h
Comment (SimonHeybrock) [inline Framework/Parallel/test/EventParserTest.h]: If I understand correctly, this is creating events in `event_id` that are sorted by `event_id` (within the pulse and bank)? Sounds like a special case, but probably unlikely to cause issues (maybe distoring performance tests?).

**Review by SimonHeybrock — COMMENTED** [review]
Some comments:

File: Framework/Parallel/inc/MantidParallel/Collectives.h
Comment (OwenArnold) [inline Framework/Parallel/inc/MantidParallel/Collectives.h]: Is there any significance of this constant? I think boost mpi has an `any_tag`

**Review by OwenArnold — COMMENTED** [review]
Looks good. I see you are still working on this though.

File: Framework/Parallel/inc/MantidParallel/Collectives.h
Comment (SimonHeybrock) [reply Framework/Parallel/inc/MantidParallel/Collectives.h]: It does (http://www.boost.org/doc/libs/1_54_0/doc/html/boost/mpi/any_tag.html), but "This constant may be used for the tag parameter of receive operations to indicate that a send with any tag will be matched by the receive." is not what we want, as far as I understand.

Comment (SimonHeybrock) [reply Framework/Parallel/inc/MantidParallel/Collectives.h]: Also this is code using the threading backend, not `boost::mpi`.

**Comment by SimonHeybrock** [comment]
I have addressed some main points I commented on earlier. There are some remaining comments I had regarding testability and tests, but they can be addressed in a follow-up PR, such that we can move forward.

@OwenArnold Do you have more comments at this point or can we merge this into the feature branch?

**Review by OwenArnold — APPROVED** [review]

**Comment by OwenArnold** [comment]
Do not merge until reviewed by @AntonPiccardoSelg

**Review by AntonPiccardoSelg — APPROVED** [review]

**Comment by OwenArnold** [comment]
As this is being merged into a long-running feature branch, and has had several reviewers, we are bypassing the separate gatekeeper. Gatekeeper will need to verify again before we can merge to master.

**Pull request merged**
